experimental/trusted_shuffler/trusted_shuffler/src/tests.rs

```diff
@@ -0,0 +1,135 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{RequestHandler, TrustedShuffler};
+use async_trait::async_trait;
+use futures::future::join_all;
+use std::{sync::Arc, time::Duration};
+use tokio::time::timeout;
+
+const TEST_TIMEOUT_MILLISECONDS: u64 = 100;
+
+struct TestRequestHandler;
+
+#[async_trait]
+impl RequestHandler for TestRequestHandler {
+    // A test function that processes requests.
+    // In the real use-case, this function should send requests to the backend and
+    // return responses.
+    async fn handle(&self, request: Vec<u8>) -> Vec<u8> {
+        generate_response(request)
+    }
+}
+
+// Non-async version of the `handle_request` function used to create expected responses.
+fn generate_response(request: Vec<u8>) -> Vec<u8> {
+    let parsed_request = String::from_utf8(request).expect("Couldn't parse request");
+    format!("Response for: {}", parsed_request)
+        .to_string()
+        .as_bytes()
+        .to_vec()
+}
+
+// Generates a request and a corresponding response from a string.
+fn generate_request_response(data: &str) -> (Vec<u8>, Vec<u8>) {
+    let request = format!("Request: {}", data).to_string().as_bytes().to_vec();
+    let response = generate_response(request.clone());
+    (request, response)
+}
+
+#[tokio::test]
+async fn trusted_shuffler_k2_test() {
+    let anonymity_value = 2;
+    let trusted_shuffler = Arc::new(TrustedShuffler::new(
+        anonymity_value,
+        Arc::new(TestRequestHandler {}),
+    ));
+
+    let (request, expected_response) = generate_request_response("Test");
+    let (background_request, expected_background_response) =
+        generate_request_response("Background test");
+
+    let trusted_shuffler_clone = trusted_shuffler.clone();
+    let background_result =
+        tokio::spawn(async move { trusted_shuffler_clone.invoke(background_request).await });
+
+    let response = trusted_shuffler.invoke(request.clone()).await;
+    assert!(response.is_ok());
+    assert_eq!(expected_response, response.unwrap());
+
+    let background_result = background_result.await;
+    assert!(background_result.is_ok());
+    let background_response = background_result.unwrap();
+    assert!(background_response.is_ok());
+    assert_eq!(expected_background_response, background_response.unwrap());
+}
+
+#[tokio::test]
+async fn trusted_shuffler_k10_test() {
+    let anonymity_value = 10;
+    let trusted_shuffler = Arc::new(TrustedShuffler::new(
+        anonymity_value,
+        Arc::new(TestRequestHandler {}),
+    ));
+
+    let (requests, expected_responses): (Vec<Vec<u8>>, Vec<Vec<u8>>) = (0..anonymity_value)
+        .collect::<Vec<_>>()
+        .iter()
+        .map(|k| generate_request_response(&format!("Test {}", k)))
+        .unzip();
+
+    let mut result_futures = vec![];
+    for request in requests.iter() {
+        let trusted_shuffler_clone = trusted_shuffler.clone();
```

Comment:
Again I'm not sure why all the Arcs and the cloning are necessary -- perhaps they are, but could you confirm that it's really the case?

Reply:
Thanks for confirming!